File: autolocalize/localization/icp_refine.py
"""Point-to-line ICP post-refinement against map wall segments."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

import numpy as np

from autolocalize.geometry.pose import Pose2D
from autolocalize.localization.config import InitialLocalizerConfig
from autolocalize.localization.fast_grid import FastOccupancyLookup
from autolocalize.map.walls import WallSegmentIndex

try:
    from autolocalize._native import WallMapNative
except ImportError:  # pragma: no cover
    WallMapNative = None  # type: ignore[misc, assignment]

if TYPE_CHECKING:
    from autolocalize._native import WallMapNative as WallMapNativeType
else:
    WallMapNativeType = Any


@dataclass(frozen=True, slots=True)
class IcpRefineResult:
    pose: Pose2D
    mean_residual: float
    iterations: int
    converged: bool


def _lookup_wall_index(
    lookup: FastOccupancyLookup,
    *,
    bucket_size: float,
) -> WallSegmentIndex | WallMapNativeType:
    cached = lookup._wall_index.get(bucket_size)
    if cached is not None:
        return cached

    if WallMapNative is not None:
        index = WallMapNative(
            lookup.occupied,
            lookup.origin_x,
            lookup.origin_y,
            lookup.resolution,
            bucket_size,
        )
    else:
        index = WallSegmentIndex.from_lookup(lookup, bucket_size=bucket_size)
    lookup._wall_index[bucket_size] = index
    return index


def refine_pose_icp(
    lookup: FastOccupancyLookup,
    pose: Pose2D,
    local_xy: np.ndarray,
    cfg: InitialLocalizerConfig,
) -> IcpRefineResult:
    """
    Optional post-localization refine: point-to-line ICP with Levenberg–Marquardt.

    Expects a pose already close to the true solution (e.g. after adaptive localize).
    """
    points = np.ascontiguousarray(local_xy, dtype=np.float64)
    if points.ndim != 2 or points.shape[1] != 2:
        raise ValueError("local_xy must be an Nx2 array")
    if points.shape[0] < cfg.icp_min_points:
        return IcpRefineResult(pose, 0.0, 0, False)

    wall_index = _lookup_wall_index(
        lookup, bucket_size=cfg.icp_bucket_size
    )

    if WallMapNative is not None and isinstance(wall_index, WallMapNative):
        x, y, theta, mean_residual, iterations, converged = wall_index.refine_icp(
            points,
            pose.x,
            pose.y,
            pose.theta,
            cfg.icp_max_iterations,
            cfg.icp_max_association_dist,
            cfg.icp_convergence_translation,
            cfg.icp_convergence_rotation,
            cfg.icp_huber_delta,
            cfg.icp_min_points,
        )
        return IcpRefineResult(
            Pose2D(x, y, theta),
            float(mean_residual),
            int(iterations),
            bool(converged),
        )

    return _refine_icp_python(
        wall_index,
        pose,
        points,
        cfg,
    )


def _refine_icp_python(
    walls: WallSegmentIndex,
    pose: Pose2D,
    local_xy: np.ndarray,
    cfg: InitialLocalizerConfig,
) -> IcpRefineResult:
    x, y, theta = pose.x, pose.y, pose.theta
    lambda_damp = 1e-3
    last_cost = math.inf
    iterations = 0
    converged = False
    mean_residual = 0.0

    for _ in range(cfg.icp_max_iterations):
        c = math.cos(theta)
        s = math.sin(theta)
        world = np.column_stack(
            (
                x + c * local_xy[:, 0] - s * local_xy[:, 1],
                y + s * local_xy[:, 0] + c * local_xy[:, 1],
            )
        )

        jtj = np.zeros((3, 3), dtype=np.float64)
        jtr = np.zeros(3, dtype=np.float64)
        cost = 0.0
        used = 0

        for i, (wx, wy) in enumerate(world):
            match = walls.nearest_segment(float(wx), float(wy), cfg.icp_max_association_dist)
            if match is None:
                continue
            residual, gx, gy = match
            abs_r = abs(residual)
            weight = (
                1.0
                if abs_r <= cfg.icp_huber_delta or cfg.icp_huber_delta <= 0.0
                else cfg.icp_huber_delta / abs_r
            )
            lx, ly = local_xy[i]
            dwx_dtheta = -s * lx - c * ly
            dwy_dtheta = c * lx - s * ly
            j = np.array([gx, gy, gx * dwx_dtheta + gy * dwy_dtheta], dtype=np.float64)
            jtj += weight * np.outer(j, j)
            jtr += weight * residual * j
            cost += weight * residual * residual
            used += 1

        iterations += 1
        mean_residual = math.sqrt(cost / used) if used else 0.0
        if used < cfg.icp_min_points:
            break

        jtj.flat[0] += lambda_damp
        jtj.flat[4] += lambda_damp
        jtj.flat[8] += lambda_damp
        try:
            delta = np.linalg.solve(jtj, -jtr)
        except np.linalg.LinAlgError:
            break

        if cost >= last_cost:
            lambda_damp = min(lambda_damp * 10.0, 1e6)
            continue

        new_x = x + float(delta[0])
        new_y = y + float(delta[1])
        new_theta = Pose2D.normalize_angle(theta + float(delta[2]))
        x, y, theta = new_x, new_y, new_theta
        lambda_damp = max(lambda_damp * 0.3, 1e-6)
        last_cost = cost

        if (
            math.hypot(float(delta[0]), float(delta[1]))
            <= cfg.icp_convergence_translation
            and abs(float(delta[2])) <= cfg.icp_convergence_rotation
        ):
            converged = True
            break

    return IcpRefineResult(Pose2D(x, y, theta), mean_residual, iterations, converged)

File: cpp/src/icp_refine.cpp
#include "icp_refine.hpp"

#include "pose_scorer.hpp"

#include <cmath>
#include <limits>
#include <vector>

namespace autolocalize {

namespace {

inline double huber_weight(double residual, double delta) {
  const double abs_r = std::abs(residual);
  if (abs_r <= delta || delta <= 0.0) {
    return 1.0;
  }
  return delta / abs_r;
}

bool solve_normal_equations(double jtj[9], double jtr[3], double lambda,
                            double delta[3]) {
  jtj[0] += lambda;
  jtj[4] += lambda;
  jtj[8] += lambda;

  const double a00 = jtj[0];
  const double a01 = jtj[1];
  const double a02 = jtj[2];
  const double a11 = jtj[4];
  const double a12 = jtj[5];
  const double a22 = jtj[8];

  double b0 = -jtr[0];
  double b1 = -jtr[1];
  double b2 = -jtr[2];

  double m00 = a00;
  double m01 = a01;
  double m02 = a02;
  double m11 = a11;
  double m12 = a12;
  double m22 = a22;

  if (std::abs(m00) < 1e-12) {
    return false;
  }
  const double inv00 = 1.0 / m00;
  m01 *= inv00;
  m02 *= inv00;
  m11 -= m01 * m01 * m00;
  m12 -= m01 * m02 * m00;
  b1 -= m01 * b0;
  b0 *= inv00;

  if (std::abs(m11) < 1e-12) {
    return false;
  }
  const double inv11 = 1.0 / m11;
  m12 *= inv11;
  m22 -= m12 * m12 * m11;
  b2 -= m12 * b1;
  b1 *= inv11;

  if (std::abs(m22) < 1e-12) {
    return false;
  }
  const double inv22 = 1.0 / m22;
  b2 *= inv22;
  b1 -= m12 * b2;
  b0 -= m01 * b1 + m02 * b2;

  delta[0] = b0;
  delta[1] = b1;
  delta[2] = b2;
  return true;
}

}  // namespace

IcpRefineResult refine_icp(const WallSegmentIndex& walls,
                           const std::vector<double>& local_xy, double x, double y,
                           double theta, int max_iterations,
                           double max_association_dist,
                           double convergence_translation,
                           double convergence_rotation, double huber_delta,
                           int min_points) {
  IcpRefineResult result{x, y, theta, 0.0, 0, false};
  const int num_points = static_cast<int>(local_xy.size() / 2);
  if (num_points < min_points || walls.segments().empty()) {
    return result;
  }

  double lambda = 1e-3;
  double last_cost = 1e300;

  for (int iter = 0; iter < max_iterations; ++iter) {
    const double c = std::cos(theta);
    const double s = std::sin(theta);

    double jtj[9] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    double jtr[3] = {0.0, 0.0, 0.0};
    double cost = 0.0;
    int used = 0;

    for (int i = 0; i < num_points; ++i) {
      const double lx = local_xy[static_cast<size_t>(2 * i)];
      const double ly = local_xy[static_cast<size_t>(2 * i + 1)];
      const double wx = x + c * lx - s * ly;
      const double wy = y + s * lx + c * ly;

      const PointSegmentDistance match =
          walls.nearest_segment(wx, wy, max_association_dist);
      if (!match.valid) {
        continue;
      }

      const double residual = match.distance;
      const double weight = huber_weight(residual, huber_delta);
      const double dwx_dtheta = -s * lx - c * ly;
      const double dwy_dtheta = c * lx - s * ly;

      const double jx = match.grad_x;
      const double jy = match.grad_y;
      const double j0 = jx;
      const double j1 = jy;
      const double j2 = jx * dwx_dtheta + jy * dwy_dtheta;

      jtj[0] += weight * j0 * j0;
      jtj[1] += weight * j0 * j1;
      jtj[2] += weight * j0 * j2;
      jtj[4] += weight * j1 * j1;
      jtj[5] += weight * j1 * j2;
      jtj[8] += weight * j2 * j2;

      const double wr = weight * residual;
      jtr[0] += j0 * wr;
      jtr[1] += j1 * wr;
      jtr[2] += j2 * wr;

      cost += weight * residual * residual;
      ++used;
    }

    result.iterations = iter + 1;
    result.mean_residual =
        used > 0 ? std::sqrt(cost / static_cast<double>(used)) : 0.0;

    if (used < min_points) {
      return result;
    }

    double delta[3] = {0.0, 0.0, 0.0};
    if (!solve_normal_equations(jtj, jtr, lambda, delta)) {
      return result;
    }

    const double new_x = x + delta[0];
    const double new_y = y + delta[1];
    const double new_theta = autolocalize::normalize_angle(theta + delta[2]);

    if (cost >= last_cost) {
      lambda = std::min(lambda * 10.0, 1e6);
      continue;
    }

    x = new_x;
    y = new_y;
    theta = new_theta;
    lambda = std::max(lambda * 0.3, 1e-6);
    last_cost = cost;

    if (std::hypot(delta[0], delta[1]) <= convergence_translation &&
        std::abs(delta[2]) <= convergence_rotation) {
      result.converged = true;
      break;
    }
  }

  result.x = x;
  result.y = y;
  result.theta = theta;
  return result;
}

}  // namespace autolocalize
